src/invariant/test/OrderBookIsNotCrossedTests.cpp

```diff
@@ -0,0 +1,333 @@
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/OrderBookIsNotCrossed.h"
+#include "invariant/test/InvariantTestUtils.h"
+#include "ledger/LedgerTxn.h"
+#include "lib/catch.hpp"
+#include "test/TestUtils.h"
+#include "test/TxTests.h"
+#include "test/test.h"
+
+using namespace stellar;
+using namespace stellar::InvariantTestUtils;
+
+// convert a single offer into an offer operation
+Operation
+opFromLedgerEntries(LedgerEntry const& le)
+{
+    auto const& offer = le.data.offer();
+    return txtest::manageBuyOffer(offer.offerID, offer.selling, offer.buying,
+                                  offer.price, offer.amount);
+}
+
+// convert a vector of offers into a path payment operation
+Operation
+opFromLedgerEntries(std::vector<LedgerEntry> const& les)
+{
+    // for given assets A, Bs, C
+    //      send: A
+    //      dest: C
+    //      path: {Bs}
+    // where Bs represents from 0 to 5 offers inclusive
+    Asset const& send = les.front().data.offer().selling;
+    Asset const& dest = les.back().data.offer().buying;
+    std::vector<Asset> path;
+
+    for (int i = 1; i < les.size(); i++)
+    {
+        path.emplace_back(les[i].data.offer().selling);
+    }
+
+    // public key and amounts are not important
+    return txtest::pathPayment(PublicKey{}, send, 1, dest, 1, path);
+}
+
+void
+applyCheck(Application& app, std::vector<LedgerEntry> const& current,
+           std::vector<LedgerEntry> const& previous, bool shouldPass,
+           Operation const& op)
+{
+    stellar::InvariantTestUtils::UpdateList updates;
+
+    if (previous.empty())
+    {
+        updates = makeUpdateList(current, nullptr);
+    }
+    else if (current.empty())
+    {
+        updates = makeUpdateList(nullptr, previous);
+    }
+    else
+    {
+        updates = makeUpdateList(current, previous);
```

Comment:
we should probably add a `REQUIRE` inside that version of `makeUpdateList` that ensures that keys are matching 1:1 (I know that code was already there, but it's super error prone especially in this context)

Reply:
Added.